core/app/models/spree/product.rb

```diff
@@ -13,6 +15,25 @@ class Product < Spree::Base
 
     acts_as_paranoid
 
+    include Discard::Model
+    self.discard_column = :deleted_at
+
+    after_discard do
+      variants_including_master.discard_all
+      self.product_option_types = []
+      self.product_properties = []
+      self.classifications = []
+      self.product_promotion_rules = []
+    end
+
+    def paranoia_destroy
+      Spree::Deprecation.warn <<-WARN.strip_heredoc
+        Calling #destroy (or #paranoia_destroy) on a Product currently performs a soft-destroy using the paranoia gem.
+        In Solidus 3.0, paranoia will be removed, and this will perform a HARD destoy instead. To continue soft-deleting, use #discard instead.
```

Comment:
"...perform a HARD _destoy_".  All three of the deprecation warnings have this.

Reply:
This also prompted me to move these warnings to a common `ParanoiaDeprecations` module.